core/trino-main/src/main/java/io/trino/sql/planner/optimizations/UnaliasSymbolReferences.java

```diff
@@ -1303,4 +1331,52 @@ public PlanNode getRoot()
             return mappings;
         }
     }
+
+    private static class DynamicFilterVisitor
+            extends SimplePlanRewriter<Void>
+    {
+        private final Metadata metadata;
+        private final Map<DynamicFilterId, DynamicFilterId> dynamicFilterIdMap;
+
+        private DynamicFilterVisitor(Metadata metadata, Map<DynamicFilterId, DynamicFilterId> dynamicFilterIdMap)
+        {
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.dynamicFilterIdMap = requireNonNull(dynamicFilterIdMap, "dynamicFilterIdMap is null");
+        }
+
+        @Override
+        public PlanNode visitFilter(FilterNode node, RewriteContext<Void> context)
+        {
+            PlanNode rewrittenSource = context.rewrite(node.getSource());
+            Expression rewrittenPredicate = updateDynamicFilterIds(dynamicFilterIdMap, node.getPredicate());
+
+            if (rewrittenSource == node.getSource() && rewrittenPredicate == node.getPredicate()) {
+                return node;
+            }
+            return new FilterNode(
+                    node.getId(),
+                    rewrittenSource,
+                    rewrittenPredicate);
+        }
+
+        private Expression updateDynamicFilterIds(Map<DynamicFilterId, DynamicFilterId> dynamicFilterIdMap, Expression predicate)
+        {
+            List<Expression> conjuncts = extractConjuncts(predicate);
+            boolean[] updated = new boolean[1]; // using array as a mutable boolean to modify it inside the lambda
```

Comment:
just use a normal loop now. It becomes too convoluted now.

Reply:
@sopel39 the updated flag has to be updated inside the Optional.map, changing this to loop will make it even more convoluted e.g.

```
        private Expression updateDynamicFilterIds(Map<DynamicFilterId, DynamicFilterId> dynamicFilterIdMap, Expression predicate)
        {
            List<Expression> conjuncts = extractConjuncts(predicate);
            boolean updated = false;
            ImmutableList.Builder<Expression> mapped = ImmutableList.builder();
            for (Expression conjunct : conjuncts) {
                Optional<DynamicFilters.Descriptor> descriptor = getDescriptor(conjunct);
                if (descriptor.isPresent()) {
                    DynamicFilterId mappedId = dynamicFilterIdMap.get(descriptor.get().getId());
                    if (mappedId != null) {
                        conjunct = replaceDynamicFilterId((FunctionCall) conjunct, mappedId);
                        updated = true;
                    }
                    mapped.add(conjunct);
                }
            }

            if (updated) {
                return combineConjuncts(metadata, mapped.build());
            }
            return predicate;
        }
```

Reply:
how about:
```
            List<Expression> conjuncts = extractConjuncts(predicate);
            boolean updated = false;
            ImmutableList.Builder<Expression> newConjuncts = ImmutableList.builder();
            for (Expression conjunct : conjuncts) {
                Optional<DynamicFilters.Descriptor> descriptor = getDescriptor(conjunct);
                if (descriptor.isEmpty()) {
                    // not DF
                    newConjuncts.add(conjunct);
                    continue; 
                }
                DynamicFilterId mappedId = dynamicFilterIdMap.get(descriptor.get().getId());
                if (mappedId != null) {
                    // DF was remapped
                    conjunct = replaceDynamicFilterId((FunctionCall) conjunct, mappedId);
                    updated = true;
                }
                newConjuncts.add(conjunct);
            }
            if (updated) {
                return combineConjuncts(metadata, newConjuncts.build());
            }
            return predicate;
```

sure it's a little more verbose, but it's straighforward. Overall, I think we should avoid stream side effects if we can.

Reply:
IMO stream version is more straightforward and less error prone but I don't want to die on this hill so I changed it to the loop